# Airline Sample Data Exercise Answer

**Q1:** How many unique customers are there? How many do not have specific age information?

In [1]:
import smv

In [2]:
# Input data
customer = openCsv("data/input/customer.csv")
member = openCsv("data/input/member_lvl.csv")
bkg = openCsv("data/input/booking.csv")
tkt = openCsv("data/input/ticket.csv")
ckin = openCsv("data/input/checkin.csv")
geo = openCsv("data/input/ref_geo.csv")

In [3]:
customer.smvEdd()

CUST_ID              Non-Null Count         22
CUST_ID              Min Length             12
CUST_ID              Max Length             12
CUST_ID              Approx Distinct Count  22
BIRTH_MM             Non-Null Count         19
BIRTH_MM             Average                6.421052631578948
BIRTH_MM             Standard Deviation     3.80519804527211
BIRTH_MM             Min                    2.0
BIRTH_MM             Max                    12.0
BIRTH_DD             Non-Null Count         19
BIRTH_DD             Average                16.263157894736842
BIRTH_DD             Standard Deviation     9.218592958817027
BIRTH_DD             Min                    1.0
BIRTH_DD             Max                    30.0
BIRTH_YYYY           Non-Null Count         19
BIRTH_YYYY           Average                1973.3157894736842
BIRTH_YYYY           Standard Deviation     11.39226166004778
BIRTH_YYYY           Min                    1953.0
BIRTH_YYYY           Max                    1991.0
GE

**A1:** 22 unique customers, 3 without age information.

**Q2:** Which customer joined FFP the earliest among all sampled customers?

In [4]:
member.filter(col("ENROLL_DT").isNotNull()).orderBy("ENROLL_DT").show(1)

+------------+--------+----------+
|     CUST_ID|CURR_LVL| ENROLL_DT|
+------------+--------+----------+
|000365709222|       5|1982-09-15|
+------------+--------+----------+
only showing top 1 row



**A2:** The customer with ID 000365709222 joined in 1982.

**Q3:** How many records are there in ticket table?

In [5]:
tkt.count()

1941

**A3:** 1941 records

One caveat is that this kind of check may take a long time on a large full data.

**Q4:** What is the missing rate of "CABIN" in booking table?

In [6]:
bkg_sample = bkg.sample(False, 0.2, 99).cache()
bkg_sample.count()

1309

In [7]:
bkg_sample.smvEdd("CABIN")

CABIN                Non-Null Count         1286
CABIN                Min Length             1
CABIN                Max Length             1
CABIN                Approx Distinct Count  4


In [8]:
float(1309-1286)/float(1309)*100

1.757066462948816

**A4:** 1.76%

**Q5:** Validate whether the set of ["CUST_ID", "PNR", "PNR_CREATION_DT", "FLT_LEG_DPRT_DT", "SCH_LEG_ORIG_CD", "SCH_LEG_DEST_CD"] can be primary keys of ticket table.

In [9]:
tkt_sample = tkt.sample(False, 0.3, 99).cache()
tkt_sample.count()

583

In [10]:
tkt_sample.smvDedupByKey("CUST_ID", "PNR", "PNR_CREATION_DT", "FLT_LEG_DPRT_DT", "SCH_LEG_ORIG_CD", "SCH_LEG_DEST_CD").count()

583

**A5:** Yes

**Q6:** What was the most frequent check-in source among sampled customers?

In [11]:
ckin.smvFreqHist("CKIN_SOURCE")

Histogram of CKIN_SOURCE: String sorted by Frequency
key                      count      Pct    cumCount   cumPct
AGENT                      586   54.77%         586   54.77%
WEB                        329   30.75%         915   85.51%
eService                   155   14.49%        1070  100.00%
-------------------------------------------------


**A6:** "AGENT" - typically means a staff checked a passenger in at the airport counter

**Q7:** At which city and which country does "AGP" locate?

In [12]:
geo.filter(col("LOCATION_ID")=="AGP").select("CITY_DESC", "COUNTRY_DESC").show()

+---------+------------+
|CITY_DESC|COUNTRY_DESC|
+---------+------------+
|   Malaga|       Spain|
+---------+------------+



**A7:** Malaga, Spain